### ISSUE TITLE
Regression: Typemap type mismatch in 1.34.0+

### ISSUE THREAD

**Issue by IslandUsurper** [rust-lang/rust#60375]
I first reported this issue at gabdube/native-windows-gui#69, but discussion on Discord led me to file an issue here.

native-windows-ui stores data about the application (form controls, state, etc.) through a typemap. While my program worked on Rust 1.32.0 and 1.33.0, it breaks on 1.34.0 and 1.34.1. Sadly, this is a runtime error, and no backtrace is generated even with the RUST_BACKTRACE flag.

The error that is generated is [`native_windows_gui::Error::BadType`](https://github.com/gabdube/native-windows-gui/search?q=BadType&unscoped_q=BadType).

**Comment by IslandUsurper** [comment]
Ping: @dpc

**Comment by Centril** [comment]
> Sadly, this is a runtime error, and no backtrace is generated even with the RUST_BACKTRACE flag.

Would it be possible to paste the runtime error?

**Comment by dpc** [comment]
@IslandUsurper Did you want to ping the other dpc? :smile:

**Comment by IslandUsurper** [comment]
@Centril, here's the error.
    thread 'main' panicked at 'Failed to find a control: They key exists in the Ui, but the type requested did not match the type of the underlying object', src\libcore\result.rs:997:5

@dpc, sorry, are you not the dpc I was talking to on Discord? Who's the other one?

**Comment by Centril** [comment]
cc @Dylan-DPC ^---

**Comment by chinuno-usami** [comment]
Same problem while running the example  in [README of natice-windows-gui](https://github.com/gabdube/native-windows-gui#code-example) .
But I got backtrace log below.
```
thread 'main' panicked at 'Failed to find a control: The key exists in the Ui, but the type requested did not match the type of the underlying object', src\libcore\result.rs:997:5
stack backtrace:
   0: std::sys::windows::backtrace::set_frames
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\sys\windows\backtrace\mod.rs:94
   1: std::sys::windows::backtrace::unwind_backtrace
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\sys\windows\backtrace\mod.rs:81
   2: std::sys_common::backtrace::_print
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\sys_common\backtrace.rs:70
   3: std::sys_common::backtrace::print
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\sys_common\backtrace.rs:58
   4: std::panicking::default_hook::{{closure}}
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\panicking.rs:200
   5: std::panicking::default_hook
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\panicking.rs:215
   6: std::panicking::rust_panic_with_hook
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\panicking.rs:478
   7: std::panicking::continue_panic_fmt
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\panicking.rs:385
   8: std::panicking::rust_begin_panic
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\panicking.rs:312
   9: core::panicking::panic_fmt
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libcore\panicking.rs:85
  10: core::result::unwrap_failed<native_windows_gui::error::Error>
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libcore\macros.rs:17
  11: core::result::Result<core::cell::Ref<alloc::boxed::Box<native_windows_gui::controls::textinput::TextInput>>, native_windows_gui::error::Error>::expect<core::cell::Ref<alloc::boxed::Box<native_windows_gui::controls::textinput::TextInput>>,native_windows_gui::error::Error>
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libcore\result.rs:825
  12: cuter_gui2::setup_ui::{{closure}}
             at .\src\main.rs:51
  13: native_windows_gui::ui::UiInner<cuter_gui2::AppId>::trigger<cuter_gui2::AppId>
             at C:\Users\cirno\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\native-windows-gui-0.2.0\src\ui.rs:359
  14: native_windows_gui::low::events::process_events<cuter_gui2::AppId>
             at C:\Users\cirno\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\native-windows-gui-0.2.0\src\low\events.rs:235
  15: Str_SetPtrW
  16: Ordinal411
  17: CallWindowProcW
  18: CallWindowProcW
  19: MapWindowPoints
  20: KiUserCallbackDispatcher
  21: NtUserMessageCall
  22: Exported
  23: SendMessageW
  24: SendMessageW
  25: Ordinal235
  26: SizeBoxHwnd
  27: CallWindowProcW
  28: CallWindowProcW
  29: Str_SetPtrW
  30: DefSubclassProc
  31: native_windows_gui::low::events::process_events<cuter_gui2::AppId>
             at C:\Users\cirno\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\native-windows-gui-0.2.0\src\low\events.rs:245
  32: Str_SetPtrW
  33: Ordinal411
  34: CallWindowProcW
  35: DispatchMessageW
  36: native_windows_gui::low::events::dispatch_events
             at C:\Users\cirno\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\native-windows-gui-0.2.0\src\low\events.rs:307
  37: native_windows_gui::ui::dispatch_events
             at C:\Users\cirno\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\native-windows-gui-0.2.0\src\ui.rs:797
  38: cuter_gui2::main
             at .\src\main.rs:75
  39: std::rt::lang_start::{{closure}}<()>
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\rt.rs:64
  40: std::rt::lang_start_internal::{{closure}}
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\rt.rs:49
  41: std::panicking::try::do_call<closure,i32>
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\panicking.rs:297
  42: panic_unwind::__rust_maybe_catch_panic
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libpanic_unwind\lib.rs:87
  43: std::panicking::try
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\panicking.rs:276
  44: std::panic::catch_unwind
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\panic.rs:388
  45: std::rt::lang_start_internal
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\rt.rs:48
  46: std::rt::lang_start<()>
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\rt.rs:64
  47: main
  48: invoke_main
             at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:64
  49: __scrt_common_main_seh
             at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:253
  50: BaseThreadInitThunk
  51: RtlUserThreadStart
```
OS : Windows 10 2016 ltsb (version 1607)
Toolchain:
stable-x86_64-pc-windows-msvc (default)
rustc 1.34.1 (fc50f328b 2019-04-24)

The last frame in main is 12 , a nwg_get! macro call, defined as 
```rust
/**
    Return controls from the Ui. Panics if one of the controls could not be retrieved.  
    
    
    To avoid the panic, use `ui.get` directly.

    Usage:  
    `let control = nwg_get!(ui; (control ID, control type))`  
    `let (control1, control2) = nwg_get!(ui; [(control1_ID, control1 type), (control2_ID, control2_type)])`  
*/
#[macro_export]
macro_rules! nwg_get {
    ( $ui:ident; ($n:expr, $t:ty) ) => {
        $ui.get::<$t>(&$n).expect("Failed to find a control")
    };

    ( $ui:ident; [ $( ($n:expr, $t:ty) ),* ] ) => {
        (
            $( $ui.get::<$t>(&$n).expect("Failed to find a control") ),*
        )
    }
}
```

**Comment by pnkfelix** [comment]
triage: Not quite clear on what is going on here; maybe something with using `TypeId` as a key in a hashmap? Leaving unprioritized and nominated for now...

**Comment by nikomatsakis** [comment]
It seems like what is really needed here is to try and narrow this to a standalone test case that is not reliant on `TypeMap` etc.

**Comment by nikomatsakis** [comment]
triage: P-high

Behavorial regression, unknown cause, needs help to track down the problem.

**Comment by gabdube** [comment]
Hey this is my project. I'll try to isolate the problem to make it a bit easier to fix... hopefully.

**Comment by gabdube** [comment]
Ok I think I've found something. Hopefully it makes sense. It seems like there are two methods named "type_id" (one from `Any` and another from my `ResourceT` trait) available and the code. The method that compare the types use the wrong one.

For example, this code does not compile:

```rust
use std::any::Any;
use std::any::TypeId;

trait Resource {
}

trait ResourceT {
    fn type_id(&self) -> TypeId;
}

struct TestResourceT {
    value: u32
}

impl ResourceT for TestResourceT {
    fn type_id(&self) -> TypeId { TypeId::of::<TestResource>() }
}

struct TestResource {}

fn main() {
    let test = TestResourceT{ value: 10 };

    println!("{:?}", TypeId::of::<TestResource>());
    println!("{:?}", test.type_id());
}
```

The compiler will output

```
   Compiling nwg-test v0.1.0 (C:\Users\Gab\Documents\projects\success\nwg-test)
error[E0034]: multiple applicable items in scope
  --> src\main.rs:32:27
   |
32 |     println!("{:?}", test.type_id());
   |                           ^^^^^^^ multiple `type_id` found
   |
note: candidate #1 is defined in an impl of the trait `ResourceT` for the type `TestResourceT`
  --> src\main.rs:18:5
   |
18 |     fn type_id(&self) -> TypeId { TypeId::of::<TestResource>() }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: candidate #2 is defined in an impl of the trait `std::any::Any` for the type `_`
```

That said, if you remove `use std::any::Any;` from the imports, everything works fine.

Here's a snippet that run the bogus code from native-windows-gui:  https://gist.github.com/gabdube/ef27e073c209b23b06d794dca9982af2

Again, removing `use std::any::Any` fix the problem.

**Comment by steveklabnik** [comment]
Triage: latest example still reproduces

**Comment by bstrie** [comment]
I reduced the failing test case:
```rust
// This program compiles but asserts when std::any::Any is in scope,
// but compiles and runs successfully when it is not in scope.
use std::any::Any;

use std::any::TypeId;

trait Resource {
    fn type_id(&self) -> TypeId;
}

impl Resource for () {
    fn type_id(&self) -> TypeId { TypeId::of::<()>() }
}

fn main() {
    let tid_orig = TypeId::of::<()>();    
    
    let ref_res: &dyn Resource = &();
    let tid_from_ref = ref_res.type_id();
    
    let box_res: Box<dyn Resource> = Box::new(());
    let tid_from_box = box_res.type_id();
    
    dbg!(tid_orig, tid_from_ref, tid_from_box);
    
    assert_eq!(tid_orig, tid_from_ref); // passes
    assert_eq!(tid_orig, tid_from_box); // fails
}

```
Take note of the asserts; the failure only happens on boxed trait objects. It took me a while to realize that `Any`'s blanket impl is only on `T: 'static`, so what's happening is that without `Any` in scope, method resolution autoderefs until it reaches `<() as Resource>::type_id`. But with `Any` in scope, it still autoderefs `&dyn Resource` (because it's not `'static`) until it gets to `<() as Resource>::type_id` (at which point I would expect the compiler to emit an ambiguous name error, and it *is* very weird to me that it doesn't do so... a possible bug?)--but for `Box<dyn Resource>`, which is `'static`, no autoderef happens and what is returned is the `TypeId` of `Box<dyn Resource>`.

So I don't think there's any action to be taken here. The behavior here is a consequence of trait objects, autoderef, and a `'static` blanket impl. 1.34 just happened to be the release when `Any::type_id` was stabilized. So while I do think that the failure to provide an error on the `&dyn Trait` case could use investigation, otherwise it seems method resolution is working as intended.

**Comment by joshtriplett** [comment]
We reviewed this in today's @rust-lang/libs meeting. Sorry that this took so long for us to get to!

This kind of issue is why we don't normally implement methods on `Box` or `Rc` or `Arc`, but `Any` has a blanket implementation for *all* static types. We talked this through, and we don't see any way around this; we'd recommend renaming the method, or calling it with the type explicitly named rather than using method syntax.
